### PR TITLE
[StaticRoute] Add preliminary support for sendfile system call.

### PR DESCRIPTION
**Pull request by jashandeep-sohi**
This is a first draft of `sendfile` support. Not merge ready, yet. Please critique.

The tests probably need some work. Should we test every static file test with both `sendfile` and `sendfile_fallback`?

https://github.com/KeepSafe/aiohttp/issues/398

### CODE REVIEW

File: aiohttp/web_urldispatcher.py [hunk 1]
```diff
@@ -145,7 +145,7 @@ class StaticRoute(Route):
 
     def __init__(self, name, prefix, directory, *,
                  expect_handler=None, chunk_size=256*1024,
-                 response_factory=None):
+                 response_factory=None, sendfile_fallback=False):
```
Comment (asvetlov) [inline aiohttp/web_urldispatcher.py]: Fallback should be performed automatically.
User doesn't know (and doesn't care) is underlying OS support sendfile.

File: aiohttp/web_urldispatcher.py [hunk 2]
```diff
@@ -163,6 +163,11 @@ def __init__(self, name, prefix, directory, *,
             raise ValueError(
                 "No directory exists at '{}'".format(self._directory))
 
+        if sendfile_fallback or not hasattr(os, "sendfile"):
+            self.sendfile = self.sendfile_fallback
+        else:
+            self.sendfile = self.sendfile_system
```
Comment (kxepal) [inline aiohttp/web_urldispatcher.py]: Better have `sendfile` method explicitly defined in class while it behaviour is controlled by just `hasattr(os, "sendfile")` - if system has sendfile support who will abuse it usage?

File: aiohttp/web_urldispatcher.py [hunk 3]
```diff
@@ -174,6 +179,63 @@ def url(self, *, filename, query=None):
...
+        try:
+            n = os.sendfile(out_fd, in_fd, offset, count)
+        except (BlockingIOError, InterruptedError):
+            return
+        except Exception as exc:
+            loop.remove_writer(out_fd)
+            fut.set_exception(exc)
+            return
+        else:
+            loop.remove_writer(out_fd)
```
Comment (asvetlov) [inline aiohttp/web_urldispatcher.py]: Please refactor code to avoid repeated remove_writer/add_writer calls.

File: aiohttp/web_urldispatcher.py [hunk 4]
```diff
@@ -174,6 +179,63 @@ def url(self, *, filename, query=None):
...
+        call.
+
+        `fobj` should be an open file object.
+
+        `resp` should be a :obj:`aiohttp.web.StreamResponse` instance.
+        """
+        yield from resp.drain()
+        req = resp._req
+        loop = req.app.loop
+        out_fd = req._transport._sock_fd
```
Comment (asvetlov) [inline aiohttp/web_urldispatcher.py]: Please use public API, e.g. `req.transport.get_extra_info('socket')`.

File: aiohttp/web_urldispatcher.py [hunk 5]
```diff
@@ -174,6 +179,63 @@ def url(self, *, filename, query=None):
...
+            loop.remove_writer(out_fd)
+
+        if n < count:
+            loop.add_writer(out_fd, self._sendfile_cb, fut, out_fd, in_fd,
+                            offset + n, count - n, loop)
+        else:
+            fut.set_result(None)
+
+    @asyncio.coroutine
+    def _sendfile(self, out_fd, in_fd, offset, count, loop):
```
Comment (kxepal) [inline aiohttp/web_urldispatcher.py]: This method doesn't looks as coroutine, then why it is?

File: aiohttp/web_urldispatcher.py (hunk 1 above)
Comment (jashandeep-sohi) [inline aiohttp/web_urldispatcher.py]: True, but it's useful to force the use of the fallback sendfile in the tests. That's the primary reason why the option exists.
But fallback is done automatically if `sendfile_fallback=False`. This should really be named `force_sendfile_fallback`, but that's too long.

File: aiohttp/web_urldispatcher.py (hunk 2 above)
Comment (jashandeep-sohi) [inline aiohttp/web_urldispatcher.py]: The main reason why I did it this way was for the unittests.

File: aiohttp/web_urldispatcher.py [hunk 6]
```diff
@@ -174,6 +179,63 @@ def url(self, *, filename, query=None):
...
+        """
+        Write the content of `fobj` to `resp` using the ``sendfile`` system
+        call.
+
+        `fobj` should be an open file object.
+
+        `resp` should be a :obj:`aiohttp.web.StreamResponse` instance.
+        """
+        yield from resp.drain()
+        req = resp._req
```
Comment (asvetlov) [inline aiohttp/web_urldispatcher.py]: Pass request object into the method, don't access private attr.

File: aiohttp/web_urldispatcher.py [hunk 7]
```diff
@@ -174,6 +179,63 @@ def url(self, *, filename, query=None):
...
+                                  loop)
+
+    @asyncio.coroutine
+    def sendfile_fallback(self, resp, fobj):
+        """
+        Mimic the :meth:`sendfile` method, but without using the ``sendfile``
+        system call. This should be used on systems that don't support
+        ``sendfile``.
+        """
+        yield from resp.drain()
```
Comment (asvetlov) [inline aiohttp/web_urldispatcher.py]: Why do you need `drain` here?

File: aiohttp/web_urldispatcher.py (hunk 1 above)
Comment (kxepal) [inline aiohttp/web_urldispatcher.py]: Public API is made not for tests, but for users. There is no need to pollute it with special flags, while you may call exact sendfile function in tests.

**Comment by asvetlov** [comment]
Yes, all tests should be passed with sendfile and fallback.
Mixin to `TestCase` can help avoid test code duplication.

File: aiohttp/web_urldispatcher.py (hunk 7 above)
Comment (jashandeep-sohi) [inline aiohttp/web_urldispatcher.py]: To flush any headers, and give the client a chance to bail out early when the chunk_size is large.

Comment (asvetlov) [inline aiohttp/web_urldispatcher.py]: That doesn't make sense: `.drain()` guarantees that data sent to OS buffer only, not derived to peer.

For sendfile version it's required, all headers should be sent to OS before pushing body.

Comment (jashandeep-sohi) [inline aiohttp/web_urldispatcher.py]: Oh, did not know that. I thought `drain` waited until the socket buffer was empty.
I'll remove it.

**Comment by asvetlov** [comment]
Thanks!

**Pull request merged**